File: lib/THNN/generic/xor.c
#ifndef TH_GENERIC_FILE
#define TH_GENERIC_FILE "generic/xor.c"
#else

void THNN_(xor_updateOutput)(
          THNNState *state,
          THTensor *input,
          THTensor *output,
          real alpha)
{
  THTensor_(resizeAs)(output, input);
  THTensor_(abs)(output, input);
  // fucking cheating
  TH_TENSOR_APPLY2(real, input, real, output,
    *output_data = fmod(*input_data, alpha);
  );
}

void THNN_(xor_updateGradInput)(
          THNNState *state,
          THTensor *input,
          THTensor *gradOutput,
          THTensor *gradInput,
          real alpha) // not used, this alpha
{
  THTensor_(resizeAs)(gradInput, input);
  TH_TENSOR_APPLY3(real, gradInput, real, gradOutput, real, input,
    *gradInput_data = *gradOutput_data;
  );
}

#endif

File: xor.lua
local xor, parent = torch.class('nn.xor', 'nn.Module')

function xor:__init(alpha)
   parent.__init(self)
   self.alpha = alpha or 1
end

function xor:updateOutput(input)
   input.THNN.xor_updateOutput(
      input:cdata(),
      self.output:cdata(),
      self.alpha
   )
   return self.output
end

function xor:updateGradInput(input, gradOutput)
   input.THNN.xor_updateGradInput(
      input:cdata(),
      gradOutput:cdata(),
      self.gradInput:cdata(),
      self.alpha
   )
   return self.gradInput
end
